File: app/src/main/java/be/digitalia/fosdem/viewmodels/EventViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.model.Event;

public class EventViewModel extends AndroidViewModel {

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final MutableLiveData<Long> eventId = new MutableLiveData<>();
	private final LiveData<Event> event = Transformations.switchMap(eventId,
			id -> {
				final MutableLiveData<Event> resultLiveData = new MutableLiveData<>();
				appDatabase.getQueryExecutor().execute(() -> {
					final Event result = appDatabase.getScheduleDao().getEvent(id);
					resultLiveData.postValue(result);
				});
				return resultLiveData;
			});

	public EventViewModel(@NonNull Application application) {
		super(application);
	}

	public boolean hasEventId() {
		return this.eventId.getValue() != null;
	}

	public void setEventId(long eventId) {
		Long newEventId = eventId;
		if (!newEventId.equals(this.eventId.getValue())) {
			this.eventId.setValue(newEventId);
		}
	}

	public LiveData<Event> getEvent() {
		return event;
	}
}

File: app/src/main/kotlin/be/digitalia/fosdem/viewmodels/EventViewModel.kt
package be.digitalia.fosdem.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.inject.CallbackViewModelAssistedFactory
import be.digitalia.fosdem.model.Event
import dev.zacsweers.metro.AppScope
import dev.zacsweers.metro.Assisted
import dev.zacsweers.metro.AssistedFactory
import dev.zacsweers.metro.AssistedInject
import dev.zacsweers.metro.ContributesIntoMap
import dev.zacsweers.metro.binding
import dev.zacsweers.metrox.viewmodel.ViewModelAssistedFactory
import dev.zacsweers.metrox.viewmodel.ViewModelAssistedFactoryKey
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async

@AssistedInject
class EventViewModel(
    @Assisted eventId: Long,
    scheduleDao: ScheduleDao,
) : ViewModel() {

    val event: Deferred<Event?> = viewModelScope.async {
        scheduleDao.getEvent(eventId)
    }

    @AssistedFactory
    @ContributesIntoMap(AppScope::class, binding = binding<ViewModelAssistedFactory>())
    @ViewModelAssistedFactoryKey(EventViewModel::class)
    fun interface Factory : CallbackViewModelAssistedFactory {
        fun create(eventId: Long): EventViewModel
    }
}
